tools/misc/cubemanip

```diff
@@ -1,280 +1,365 @@
+expression are substituted by the data arrays of the appropriate cube files. The
+resulting cube file will have the same header as the first input file, while its
+data array is created by evaluating the expression. The number of cube files
+passed as arguments must be higher than the maximal reference index in the
+expression.
+
+Apart of the data arrays, following variables and functions can be used in the
+expression:
+
+  origin:  origin of the grid in the first input file as (3,) shaped array
+
+  gridvecs: grid vectors in the first input file as (3, 3) shaped array
+
+  ngrids: nr. of grid units along the grid vectors as a (3,) shaped array
+
+  posx(), posy(), posz(): x-, y- and z- coordinates of each grid point
+
+Examples:
+
+  Weighted sum of three cube files:
+  {SCRIPT_NAME} '0.25 * ($1) + 0.25 * ($2) + 0.5 * ($3)' f1.cube f2.cube f3.cube
+
+  Integral of the values in a cube file:
+  {SCRIPT_NAME} -d 'sum(($1))' file1.cube
+
+  Squared norm of the data in a cube file:
+  {SCRIPT_NAME} -d 'sum(($1)**2) * det(gridvecs)' file1.cube
+"""
+
+
+@dataclass
+class Cube:
+    """Contains data found in a cube file
+
+    Attributes:
+        species: Integer species index for each atom, shape: (natoms,).
+        coords: Cartesian coordinates of each atom, shape: (natoms, 3).
+        origin: Origin of the grid, shape: (3, 3).
+        gridvecs: Vectors defining shifts between neighbouring grid points, shape: (3, 3).
+        ngrids: Nr. of grid points along each grid vector, shape (3,).
+        data: Volumentric data for each grid point, with shape of ngrids.
+        natoms: Nr. of atoms in the system.
+    """
+
+    species: np.ndarray
+    coords: np.ndarray
+    origin: np.ndarray
+    gridvecs: np.ndarray
+    ngrids: np.ndarray
+    data: np.ndarray
+    natoms: int = field(default=0, init=False)
+
+
+    def __post_init__(self):
+        """Cheks array compatibilities, fills up uninitialized fields."""
+
+        if len(self.coords.shape) != 2 or self.coords.shape[1] != 3:
+            raise ValueError(_invalid_shape_msg("coords", (-1, 3), self.coords.shape))
+        self.natoms = len(self.coords)
+
+        if self.species.shape != (self.natoms,):
+            raise ValueError(_invalid_shape_msg("species", (self.natoms,), self.species.shape))
+
+        if self.origin.shape != (3,):
+            raise ValueError(_invalid_shape_msg("origin", (3,), self.origin.shape))
+
+        if self.gridvecs.shape != (3, 3):
+            raise ValueError(_invalid_shape_msg("gridvecs", (3, 3), self.gridvecs.shape))
+
+        if self.ngrids.shape != (3,):
+            raise ValueError(_invalid_shape_msg("ngrids", (3,), self.ngrids.shape))
+
+        if self.data.shape != tuple(self.ngrids):
+            raise ValueError(_invalid_shape_msg("data", self.ngrids, self.data.shape))
+
+
+    @classmethod
+    def from_file(cls, fname):
+        """Creates a Cube instance from a file.
```

Comment:
```suggestion
        """Creates a cube instance from a file.
```
Or is this a proper name?

Reply:
It is the name of the class, which is capitalized. :wink: